### PR TITLE
LiquidityGauge v2

### PR DESCRIPTION
**Pull request by iamdefinitelyahuman**
### What I did
An update to the liquidity gauge contract. Features include:

* tokenization of deposits, allowing LP tokens to be transferred without withdrawal
* generalized staking ability in any unipool-style contract
* distribution of multiple reward tokens
* disabling or modifying the active staking contract address

### CODE REVIEW

File: contracts/gauges/LiquidityGaugeV2.vy [hunk 1]
```diff
@@ -0,0 +1,720 @@
...
+    @param _reward_tokens Array of reward token addresses to claim
+    @param _addr Address to claim for
+    """
+    for token in _reward_tokens:
+        if token == ZERO_ADDRESS:
+            break
+        integral: uint256 = self.reward_integral[token]
+        integral_for: uint256 = self.reward_integral_for[token][_addr]
+
+        if integral_for < integral:
```
Comment (michwill) [inline contracts/gauges/LiquidityGaugeV2.vy:393]: Can it ever be >?
What if it's `==`? For just one user?

Comment (iamdefinitelyahuman) [reply contracts/gauges/LiquidityGaugeV2.vy:393]: `>` should not be possible. `==` would be if the user tries to claim twice in the same block, or claims multiple times on a token where rewards have finished.

File: contracts/gauges/LiquidityGaugeV2.vy [hunk 2]
```diff
@@ -0,0 +1,735 @@
...
+        sigs: bytes32 = _sigs
+        deposit_sig: Bytes[4] = slice(sigs, 0, 4)
+        withdraw_sig: Bytes[4] = slice(sigs, 4, 4)
+
+        if convert(deposit_sig, uint256) != 0:
+            # need a non-zero total supply to verify the sigs
+            assert total_supply != 0  # dev: zero total supply
+            ERC20(lp_token).approve(_reward_contract, MAX_UINT256)
+
+            # it would be Very Bad if we get the signatures wrong here, so
```
Comment (michwill) [inline contracts/gauges/LiquidityGaugeV2.vy:669]: Whoah, loving the fool-proof approach

**Review by michwill — APPROVED** [review]

**Comment by Ronnygiron1** [comment]
Please be advised my assets, along with my financial support, and credit is being forged, I am in process of retracting, and abstracting, I am Ronny Alexander Giron gironronny1@gmail.com

**Pull request merged**